src/components/graph/Graph.jsx

```diff
@@ -302,6 +307,12 @@ export default class Graph extends React.Component {
         d3SelectAll(`#${this.state.id}-${CONST.GRAPH_CONTAINER_ID}`).attr("transform", transform);
 
         this.state.config.panAndZoom && this.setState({ transform: transform.k });
+
+        // only send zoom change events if the zoom has changed (_zoomed() also gets called when panning)
+        if (this.props.onZoomChange && this.state.previousZoom !== transform.k) {
+            this.setState({ previousZoom: transform.k });
```

Comment:
Since we're adding a new API, let's make sure we cover as much use cases as possible; for that, I would ask you to also add the `currentZoom`, as in:

```javascript
this.props.onZoomChange(this.state.previousZoom, transform.k);
```

Reply:
I feel a bit preoccupied with calling this here without any debouncing mechanism. This callback will execute potentially dozens of times per each zoom interaction.

Here's an implementation of `debounce`, you can add it to the `src/utils.js`.

```javascript
function debounce(fn, time) {
  let timer;

  return function exec(...args) {
    const later = () => {
        clearTimeout(timer);
        fn(...args);
    };
    timer && clearTimeout(timer);
    timer = setTimeout(later, wait);
  };
};
```

Then you'll need to hold a value of a denounced version of the callback in the Graph component, so in the `UNSAFE_componentWillReceiveProps` you'll have to set its value:

```javascript
if (this.props.onZoomChange) {
  this.debouncedOnZoomChange = debounce(this.props.onZoomChange, 50);
}
```

Not too sure of the `50ms` maybe we can push it to `100ms`? Give it a try and see how it feels. But it will have a great impact in performance, if we don't do it I foresee that people using this feature will get a sluggish graph.

Reply:
Debouncing is a great idea; I did notice that this is called a _lot_ of times the way it is right now!